File: packages/sdk-ts/src/browserClipboard.ts
import type { ContextClipboardPasteParams } from "@browserbasehq/stagehand-protocol/types";
import { StagehandMethods } from "@browserbasehq/stagehand-protocol/schema-registry";
import type { Page } from "./page.js";
import type { StagehandCommandClient } from "./commandClient.js";

export type ClipboardOptions = {
  page?: Page;
};

export type ClipboardPasteOptions = ClipboardOptions & {
  shortcut?: ContextClipboardPasteParams["shortcut"];
};

export class BrowserClipboard {
  constructor(readonly rpcClient: StagehandCommandClient) {}

  async readText(options?: ClipboardOptions): Promise<string> {
    return await this.rpcClient.send(
      StagehandMethods.contextClipboardReadText,
      clipboardTarget(options),
    );
  }

  async writeText(text: string, options?: ClipboardOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.contextClipboardWriteText, {
      text,
      ...clipboardTarget(options),
    });
  }

  async clear(options?: ClipboardOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.contextClipboardClear, clipboardTarget(options));
  }

  async paste(options?: ClipboardPasteOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.contextClipboardPaste, {
      ...clipboardTarget(options),
      ...(options?.shortcut === undefined ? {} : { shortcut: options.shortcut }),
    });
  }

  async copy(options?: ClipboardOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.contextClipboardCopy, clipboardTarget(options));
  }

  async cut(options?: ClipboardOptions): Promise<void> {
    await this.rpcClient.send(StagehandMethods.contextClipboardCut, clipboardTarget(options));
  }
}

function clipboardTarget(options?: ClipboardOptions): { pageId?: string } {
  return options?.page ? { pageId: options.page.pageId } : {};
}

File: packages/sdk-go/browser_clipboard.go
package stagehand

import "context"

// ClipboardOptions optionally scopes a clipboard operation to one page.
type ClipboardOptions struct {
	Page *Page
}

// ClipboardPasteOptions optionally scopes paste and selects its shortcut.
type ClipboardPasteOptions struct {
	Page     *Page
	Shortcut *ContextClipboardPasteParamsShortcut
}

// BrowserClipboard exposes context clipboard operations.
type BrowserClipboard struct {
	rpc protocolClient
}

// ReadText returns the current clipboard text.
func (c *BrowserClipboard) ReadText(ctx context.Context, options *ClipboardOptions) (string, error) {
	params := clipboardTarget(options)
	var result ContextClipboardReadTextResult
	if err := c.rpc.call(ctx, "context.clipboard_read_text", params, &result); err != nil {
		return "", err
	}
	return string(result), nil
}

// WriteText replaces the current clipboard text.
func (c *BrowserClipboard) WriteText(ctx context.Context, text string, options *ClipboardOptions) error {
	params := ContextClipboardWriteTextParams{PageID: pageID(options), Text: text}
	var result ContextVoidResult
	return c.rpc.call(ctx, "context.clipboard_write_text", params, &result)
}

// Clear clears the clipboard.
func (c *BrowserClipboard) Clear(ctx context.Context, options *ClipboardOptions) error {
	params := clipboardTarget(options)
	var result ContextVoidResult
	return c.rpc.call(ctx, "context.clipboard_clear", params, &result)
}

// Paste pastes the clipboard into the active element.
func (c *BrowserClipboard) Paste(ctx context.Context, options *ClipboardPasteOptions) error {
	params := ContextClipboardPasteParams{}
	if options != nil {
		params.PageID = pageID(&ClipboardOptions{Page: options.Page})
		params.Shortcut = options.Shortcut
	}
	var result ContextVoidResult
	return c.rpc.call(ctx, "context.clipboard_paste", params, &result)
}

// Copy copies the current selection.
func (c *BrowserClipboard) Copy(ctx context.Context, options *ClipboardOptions) error {
	params := clipboardTarget(options)
	var result ContextVoidResult
	return c.rpc.call(ctx, "context.clipboard_copy", params, &result)
}

// Cut cuts the current selection.
func (c *BrowserClipboard) Cut(ctx context.Context, options *ClipboardOptions) error {
	params := clipboardTarget(options)
	var result ContextVoidResult
	return c.rpc.call(ctx, "context.clipboard_cut", params, &result)
}

func clipboardTarget(options *ClipboardOptions) ContextClipboardTarget {
	return ContextClipboardTarget{PageID: pageID(options)}
}

func pageID(options *ClipboardOptions) *string {
	if options == nil || options.Page == nil {
		return nil
	}
	pageID := options.Page.PageID()
	return &pageID
}
